### PR TITLE
feat: Plugin pre post cd

### PR DESCRIPTION
**Pull request by milindk8**
# Description

Added plugin support in CDpipeline creation 

Fixes [#AB3778](https://dev.azure.com/DevtronLabs/Devtron/_boards/board/t/Infra/Stories/?workitem=3778)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/CIPipelineN/Sidebar.tsx [hunk 1]
```diff
@@ -102,13 +89,73 @@ export function Sidebar({
...
+        setFormData(_form)
+    }
+
+
+    const renderConfigSecret = () => {
+
+        const preStageValue = [...formData.preStageConfigMapSecretNames.configMaps, ...formData.preStageConfigMapSecretNames.secrets]
+        const postStageValue = [...formData.postStageConfigMapSecretNames.configMaps,...formData.postStageConfigMapSecretNames.secrets]
+        const valueList = activeStageName === BuildStageVariable.PreBuild ? preStageValue : postStageValue
+        console.log(valueList);
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/Sidebar.tsx:129]: Please remove consol.log

File: src/components/CIPipelineN/Sidebar.tsx [hunk 2]
```diff
@@ -160,6 +207,7 @@ export function Sidebar({
                     </div>
                 </div>
             )}
+            {activeStageName !== BuildStageVariable.Build && renderConfigSecret()}
```
Comment (vivek-devtron) [inline src/components/CIPipelineN/Sidebar.tsx:210]: Will it be available for CI as well?

File: src/components/cdPipeline/BuildCD.tsx [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+        })
+        _form.savedStrategies = strategies
+        setFormData(_form)
+    }
+
+    const renderStrategyOptions = () => {
+        return (
+            <Select
+                rootClassName="deployment-strategy-dropdown br-0 bw-0 w-150"
+                onChange={(e) => selectStrategy(e.target.value)}
```
Comment (vivek-devtron) [inline src/components/cdPipeline/BuildCD.tsx:456]: PLease avoid inline function

File: src/components/cdPipeline/BuildCD.tsx [hunk 4]
```diff
@@ -0,0 +1,623 @@
...
+    return pageState === ViewType.LOADING.toString() ? (
+        <div style={{ minHeight: '200px' }} className="flex">
+            <Progressing pageLoader />
+        </div>
+    ) : (
+        <div className="p-20 ci-scrollable-content">
+            {isAdvanced && renderPipelineNameInput()}
+            <p className="fs-14 fw-6 cn-9">Deploy to environment</p>
+            {renderEnvNamespaceAndTriggerType()}
+            {!window._env_.HIDE_GITOPS_OR_HELM_OPTION && !isVirtualEnvironment && renderDeploymentAppType()}
```
Comment (vivek-devtron) [inline src/components/cdPipeline/BuildCD.tsx:619]: We can move these conditions inside the method

File: src/components/cdPipeline/NewCDPipeline.tsx [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+                        label: secret,
+                        value: secret,
+                        type: 'secrets'
+                    }
+                })
+                : [],
+        }    
+    }
+
+    const validateTask = (taskData: StepType, taskErrorObj: TaskErrorObj): void => {
```
Comment (vivek-devtron) [inline src/components/cdPipeline/NewCDPipeline.tsx:356]: If possible can we extract the common code to some utils file?

File: src/components/cdPipeline/NewCDPipeline.tsx [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+        const _formDataErrorObj = { ...formDataErrorObj }
+        _formDataErrorObj[activeStageName].steps.push({
+            name: { isValid: true, message: null },
+            isValid: true,
+        })
+        setFormDataErrorObj(_formDataErrorObj)
+        setSelectedTaskIndex(_formData[activeStageName].steps.length - 1)
+    }
+
+    const calculateLastStepDetail = (
```
Comment (vivek-devtron) [inline src/components/cdPipeline/NewCDPipeline.tsx:488]: Same here

File: src/components/cdPipeline/NewCDPipeline.tsx [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+        newSelection['jsonStr'] = JSON.stringify(allStrategies.current[value], null, 4)
+        newSelection['yamlStr'] = yamlJsParser.stringify(allStrategies.current[value], { indent: 2 })
+
+        const _form = {...formData}
+        _form.savedStrategies.push(newSelection)
+        _form.savedStrategies = [newSelection]
+        setFormData(_form)
+    }
+
+    const validateStage = (stageName: string, _formData: CDFormType, formDataErrorObject?): void => {
```
Comment (vivek-devtron) [inline src/components/cdPipeline/NewCDPipeline.tsx:611]: Same here

File: src/components/cdPipeline/NewCDPipeline.tsx [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+        const _formDataErrorObj = {
+            ...(formDataErrorObject ?? formDataErrorObj),
+            name: validationRules.name(_formData.name),
+        } // validating name always as it's a mandatory field
+        if (stageName === BuildStageVariable.Build) {
+            _formDataErrorObj[BuildStageVariable.Build].isValid = _formDataErrorObj.name.isValid
+
+            let valid = _formData.materials.reduce((isValid, mat) => {
```
Comment (vivek-devtron) [inline src/components/cdPipeline/NewCDPipeline.tsx:619]: Will this be required in CD?

File: src/components/workflowEditor/workflowEditor.tsx [hunk 9]
```diff
@@ -329,15 +332,14 @@ class WorkflowEdit extends Component<WorkflowEditProps, WorkflowEditState> {
                     <Route
                         path={[URLS.APP_LINKED_CI_CONFIG, URLS.APP_CI_CONFIG, PipelineType.WEBHOOK].map(
                             (pipeline) =>
-                                `${this.props.match.path}/${pipeline}/:ciPipelineId/cd-pipeline/:cdPipelineId?`,
+                                `${this.props.match.path}/${pipeline}/:ciPipelineId/cd-pipeline/:cdPipelineId`,
                         )}
-                        render={({ location, history, match }: { location: any; history: any; match: any }) => {
+                        render={({ location, match }: { location: any; match: any }) => {
```
Comment (vivek-devtron) [inline src/components/workflowEditor/workflowEditor.tsx:337]: We can remove the render metjod and simplify it like child component

Comment (milindk8) [reply src/components/workflowEditor/workflowEditor.tsx:336]: We are doing some computation and sending props in render method

File: src/components/workflowEditor/workflowEditor.tsx [hunk 10]
```diff
@@ -34,7 +33,7 @@ import DeprecatedWarningModal from './DeprecatedWarningModal'
 import nojobs from '../../assets/img/empty-joblist@2x.png'
 import NewCDPipeline from '../cdPipeline/NewCDPipeline'
 
-export const pipelineContext = createContext(null)
+export const pipelineContext = createContext<PipelineContext<any>>(null)
```
Comment (vivek-devtron) [inline src/components/workflowEditor/workflowEditor.tsx:36]: Why are we passing any here?

File: src/components/cdPipeline/NewCDPipeline.tsx [hunk 11]
```diff
@@ -878,78 +878,72 @@ export default function NewCDPipeline({
...
-                            pageState,
-                            setPageState,
-                            globalVariables,
-                            configMapAndSecrets,
-                            getPrePostStageInEnv,
-                            isVirtualEnvironment,
-                            setInputVariablesListFromPrevStep
-                        }}
+                <pipelineContext.Provider
+                    value={{
```
Comment (vivek-devtron) [inline src/components/cdPipeline/NewCDPipeline.tsx:882]: Create memoised variable for context values

File: src/components/cdPipeline/NewCDPipeline.tsx [hunk 12]
```diff
@@ -0,0 +1,1064 @@
...
+            _formDataErrorObj.nameSpaceError = validationRules.namespace(_formData.namespace)
+        }
+        if(formData.generatedHelmPushAction === GeneratedHelmPush.PUSH){
+            _formDataErrorObj.containerRegistryError = validationRules.containerRegistry(
+                formData.containerRegistryName || '',
+            )
+            _formDataErrorObj.repositoryError = validationRules.repository(formData.repoName)
+        }
+
+        // if (formData.generatedHelmPushAction === GeneratedHelmPush.PUSH) {
```
Comment (vivek-devtron) [inline src/components/cdPipeline/NewCDPipeline.tsx:634]: Why this is commented?

File: src/components/cdPipeline/NewCDPipeline.tsx [hunk 13]
```diff
@@ -0,0 +1,1064 @@
...
+                    title={`Delete '${formData.name}' ?`}
+                    description={`Are you sure you want to delete this CD Pipeline from '${appName}' ?`}
+                    delete={() => handleDeletePipeline(DELETE_ACTION.DELETE)}
+                    closeDelete={hideDeleteModal}
+                />
+            )
+        }
+    }
+
+    const getNavLink = (toLink: string, stageName: string) => {
```
Comment (vivek-devtron) [inline src/components/cdPipeline/NewCDPipeline.tsx:841]: We should move this and other similar methods to utils so that it can be reused

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**